app/src/main/java/org/ea/sqrl/processors/CPSServer.java

```diff
@@ -0,0 +1,257 @@
+package org.ea.sqrl.processors;
+
+import android.content.Context;
+import android.util.Log;
+
+import org.ea.sqrl.R;
+import org.ea.sqrl.utils.EncryptionUtils;
+import org.ea.sqrl.utils.Utils;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Provides a minimalistic "web server" on port 25519 to support
+ * SQRL's "Client Provided Session (CPS)" authentication protection mechanism.
+ *
+ */
+public class CPSServer {
+    private static final String TAG = "CPSServer";
+    private static final int CPS_SERVER_PORT = 25519;
+
+    private static CPSServer mInstance = null;
+    private static CommunicationFlowHandler mCommFlowHandler = null;
+    private static Context mContext;
+    private ServerSocket mServerSocket;
+    private boolean mSentImage = false;
+    private Thread mCpsThread;
+    private boolean mCancelCPS = false;
+
+    private CPSServer() {}
+
+    public static CPSServer getInstance(Context context, CommunicationFlowHandler communicationFlowHandler) {
+        mContext = context;
+        mCommFlowHandler = communicationFlowHandler;
+
+        if(mInstance == null) {
+            mInstance = new CPSServer();
+        }
+
+        return mInstance;
+    }
+
+    public boolean start(Runnable doneAction) {
+        mCpsThread = new Thread(() -> {
+
+            mSentImage = false;
+            boolean done = false;
+
+            try {
+                mServerSocket = new ServerSocket(CPS_SERVER_PORT);
+
+                while (!mServerSocket.isClosed() && !done) {
+                    Socket socket = mServerSocket.accept();
+                    BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+
+                    String requestLine = in.readLine();
+                    Log.i(TAG, requestLine);
+
+                    Map<String, String> headers = parseHeaders(in);
```

Comment:
Do we really need to create a map with all the headers in order to look for one? Do we want to use the headers for other work?

Reply:
Only searching for "origin:" within the whole string would be too much of a messy hack in my opinion. So we'd have to iterate over the headers anyway, since we don't know the exact position of the wanted header. So why not do it properly in the first place?

I prefer writing readable, self-documenting and maintainable code over over-optimizing. I once read in a paper one should write nice code and only start optimizing if there are actual performance-problems. This struck a chord with me.

I would even go so far as to say that all the ad-hoc HTTP request parsing code which is present in multiple places within the project at the moment should be broken out into a common request parsing class. This would vastly increase maintainability and readability of the codebase.

But that's only my personal preference, and I have no problem at all with following the project lead when it comes to questions like that. So if you have a concrete suggestion on how to improve this part of the code, please let me know and I'll implement it that way! 😁

Reply:
Thanks for the lengthy comment and I think there is something to the point. I've optimized code to early many times and I've worked with memory constrained environments.

And much of the optimizations can be handled by the compiler. So for this code it does not matter but for the future we need to keep in mind that I want the project to work on ANY phone that uses API 16 and above.

We want a broad reach so can we constrain the memory usage we should. But I'm not one to argue over a few bytes :)

Reply:
Great discussion, and point taken on the topic of keeping memory usage for older device down! :)